data/window.go

```diff
@@ -0,0 +1,168 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package data
+
+import (
+	"context"
+	"sort"
+	"sync"
+	"time"
+)
+
+type DataPoint struct {
+	Timestamp time.Time
+	Count     uint32
+}
+
+type slidingWindow struct {
+	resolution       time.Duration
+	length           time.Duration
+	evictionInterval time.Duration
+	mutex            sync.Mutex
+	dataPoints       []DataPoint
+}
+
+func NewSlidingWindow(resolution, length, evictionInterval time.Duration) *slidingWindow {
+	s := &slidingWindow{
+		resolution:       resolution,
+		length:           length,
+		evictionInterval: evictionInterval,
+		dataPoints:       []DataPoint{},
+	}
+
+	s.consolidate()
+
+	return s
+}
+
+func (s *slidingWindow) Start(ctx context.Context) {
+	done := ctx.Done()
+
+	go func() {
+		ticker := time.NewTicker(s.evictionInterval)
+
+		for {
+			select {
+			case <-done:
+				return
+			case <-ticker.C:
+				s.mutex.Lock()
+				s.slide()
+				s.consolidate()
+				s.mutex.Unlock()
+			}
+		}
+	}()
+}
+
+// maxLength calculates the maximum number of dataPoints in the sliding window
+func (s *slidingWindow) maxLength() int {
+	return int(s.length / s.resolution)
+}
+
+// slide removes dataPoints older than length
+func (s *slidingWindow) slide() {
+	now := time.Now().Truncate(s.resolution)
+
+	cutoff := -1
+	for j := len(s.dataPoints) - 1; j >= 0; j-- {
+		if s.dataPoints[j].Timestamp.Before(now.Add(-s.length)) {
+			cutoff = j + 1
+			break
+		}
+	}
+
+	if cutoff == -1 {
+		return
+	}
+
+	s.dataPoints = s.dataPoints[cutoff:]
+}
+
+// AddCount adds +1 to the DataPoint at the correct moment
+func (s *slidingWindow) AddCount(at time.Time) {
+	// first we truncate the at time to the correct resolution
+	at = at.Truncate(s.resolution)
+
+	// then we check if the at time is in the sliding window
+	if at.Before(time.Now().Add(-s.length).Truncate(s.resolution)) {
+		// the at time is before the sliding window, we ignore it
+		return
+	}
+
+	// add the Count to the correct DataPoint
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	for i, dataPoint := range s.dataPoints {
+		if dataPoint.Timestamp.Equal(at) {
+			s.dataPoints[i].Count++
+			return
+		}
+	}
+
+	// no DataPoint found, create a new one
+	s.dataPoints = append(s.dataPoints, DataPoint{
+		Timestamp: at,
+		Count:     1,
+	})
+
+	// consolidate the window
+	s.consolidate()
+
+	return
+}
+
+// consolidate the window, this will fill any gaps in the dataPoints slice
+// it'll first truncate the slice and remove all old dataPoints
+// then it'll fill the slice with new DataPoints
+// and sort it afterwards
+func (s *slidingWindow) consolidate() {
```

Comment:
This also drops any future datapoint added by AddCount. Due to clock drift between systems we might miss some of the transactions in what the monitor still considers the next interval. The impact this has depends on the size of s.resolution.
Adding 1 (hidden) bucket/interval in the future will solve this. Not sure if it is worth the complexity?

Reply:
I'll apply a clockdift of 5 secs to each data point so the newest won't fall off, but maybe the oldest when sliding.